### DISCUSSION TITLE
MathQuill equation input unexpected behavior.

### DISCUSSION THREAD

**Discussion by hiiambob89** [scenerystack/community#38] [Q&A] [answered]
We got Mathquill working in a custom component, but there were issues with registering clicks into the Mathquill components, but we got it fixed with custom logic. However, it is not able to detect where in the equation to put the cursor (Besides on firefox). Is there anyway to deal with this?
[Here](https://edwardslabusu.github.io/DynamicsLab/bead_on_hoop/bead-on-hoop_en_adapted-from-phet.html) is the simulation hosted.
Here is the custom component:
```js
export default class EquationInput extends Node {
  constructor( screenNode,parentDomElement, options ) {

    options = _.merge( {

      // can limit contents of the input
      inputType: 'text', // 'text' | 'number'

      domElement: null,

      onlyLetters: false,

      // number of characters if you want to customize
      width: null,

      // the initial text for the element
      initialText: '',

      equation: null,
      
      verifyEq: null,
      dom:null,
      // errField: null,

      linkObj: null,

      focusedEq: null

    }, options );

    super( options );

    this.valueSubmittedEmitter = new phet.axon.Emitter();
    this.writeToEquation = function(text){
      this.mathField.write(text)
      // options.model[options.equation]= mathField.latex().replaceAll('\\theta',' t').replaceAll('\\omega',' o');
      // options.model.validEqs = options.verifyEq(domElement, options.model);

    }
    const domElement = parentDomElement; 
    this.domElement = options.dom;
    const mathField = window.MathQuill.getInterface(2).MathField(domElement, {
    // console.log(getInterface)
    // console.log(MathQuill)
    // const mathField = MathQuill.getInterface(2).MathField(domElement, {
      handlers: {
        edit: function() {
          // console.log("balkjsdflajsdbfaklsjdf")
          options.model[options.equation]= mathField.latex().replaceAll('\\theta',' t').replaceAll('\\omega',' o');
          // console.log(mathField.latex())
          options.model.validEqs = options.verifyEq(domElement, options.model);
        },

      }
    });
    // console.log("hereee",mathField)
    this.mathField = mathField;
    mathField.write(options.initialText);

    const equationInputNode = new Node();
    this.equationInputNode = domElement;
    const domElementDOM = new phet.scenery.DOM( domElement ) ;
    equationInputNode.addChild( domElementDOM);
    this.addChild(equationInputNode);
    

    domElement.addEventListener( 'keyup', event => {
      if ( event.code === phet.scenery.KeyboardUtils.KEY_ENTER ) {

        this.valueSubmittedEmitter.emit( mathField.latex() ); 

        mathField.blur();
      }
    } );


    this.domElement = domElement;
  }

  getElementValue() {
    return this.mathField.latex()
  }

  
}
`
Click registration and node wrapping example:
`
    const velocityField = new EquationInput(this.sceneNode, velocityDOM,
      {
        initialText: model.velocityEQ,
        model: model,
        equation: 'velocityEQ',
        verifyEq: verifyEq,
        dom: velocityDOM,
        focusedEq: this.focusedEq
      }


    );
        this.focusedEq = [model.velocityEQ, velocityField, velocityField.writeToEquation];
    const focusedEq = [this.focusedEq]
    const velBox = new HBox({ preferredWidth: 30, children: [velocityField] })
        velBox.addInputListener({ 

      down: function (event) {
        velocityField.mathField.focus();
        focusedEq[0][0] = model.velocityEQ
        focusedEq[0][1] = velocityField

        console.log('Click inside velocity equation detected');
      }
    });
```

**Comment by hiiambob89** [comment]
Apologies for the formatting getting messed up

**Reply by zepumph** [reply]
No worries! I hope it's ok that I edited you comment, changing your single backticks into triple backtics: 

    ```js
    
    . . . .
    
    ```

**Comment by jonathanolson** [comment]
I'm looking into this. I recently experienced something similar (we don't usually embed text elements in a phet.scenery.DOM, but I was trying to), and my commit from a week ago (https://github.com/phetsims/scenery/commit/323ab8fe565c27c4442162042d41096991b7050f) somewhat improved this. If you are using code that is up-to-date, could you try providing the option `{ allowInput: true }` to the phet.scenery.DOM? e.g.:

```js
const domElementDOM = new phet.scenery.DOM( domElement, { allowInput: true } ) ;
```

By default, we call preventDefault() on certain types of events. This turns off that behavior for some events, if it detects the event is related to the DOM node's domElement.

There are also a few options passed to the Scenery Display which may be helpful (but which our common code doesn't currently expose easily, as it looks like Joist is being used). Sim creates a SimDisplay, and if you pass arguments to the SimDisplay's super call somehow, the following options may be useful (may work, may not):

- `passiveEvents: true` - Will add the `passive` flag to DOM event listeners, and we won't call preventDefault() on a larger range of things. May cause other input issues
- `allowCSSHacks: false` - In our simulations, we apply some invasive CSS rules by default that can affect interaction. We do this to avoid users accidentally "selecting" SVG/Canvas/etc., but things like CSS userDrag/userSelect/touchAction/onselectstart overrides (noted in https://github.com/phetsims/scenery/blob/c2ab243070e61bf9778ca0afabd0169f15e05839/js/display/Display.ts#L1153-L1169) may be interfering.

I would like to have text input working smoothly! Let me know if any of these help. Feel free to reach out, and I'll be independently taking a better look at text input.

**Reply by jonathanolson** [reply, marked as answer]
Looks like the Display options are likely not needed. I'd recommend trying the above `allowInput: true`, along with making sure you have the following commit: https://github.com/phetsims/scenery/commit/7603e7061576c8f88324d108b1c87ba13aa76ee9

Turns out we were calling setPointerCapture on pointerdown events, which was messing with text input a bunch.

**Reply by hiiambob89** [reply]
Ok, just tried it out, it seems to work, thanks!
